storage/replica_command.go

```diff
@@ -229,43 +230,63 @@ func (r *Replica) Scan(batch engine.Engine, args proto.ScanRequest) (proto.ScanR
 
 // EndTransaction either commits or aborts (rolls back) an extant
 // transaction according to the args.Commit parameter.
-func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args proto.EndTransactionRequest) (proto.EndTransactionResponse, error) {
+func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args proto.EndTransactionRequest) (proto.EndTransactionResponse, []proto.Intent, error) {
 	var reply proto.EndTransactionResponse
 
 	if args.Txn == nil {
-		return reply, util.Errorf("no transaction specified to EndTransaction")
+		return reply, nil, util.Errorf("no transaction specified to EndTransaction")
 	}
 	// Make a copy of the transaction in case it's mutated below.
 	txn := *args.Txn
 	key := keys.TransactionKey(txn.Key, txn.ID)
 
 	// Fetch existing transaction if possible.
-	existTxn := &proto.Transaction{}
-	ok, err := engine.MVCCGetProto(batch, key, proto.ZeroTimestamp, true, nil, existTxn)
+	reply.Txn = &proto.Transaction{}
+	ok, err := engine.MVCCGetProto(batch, key, proto.ZeroTimestamp, true, nil, reply.Txn)
 	if err != nil {
-		return reply, err
+		return reply, nil, err
+	}
+	if !ok {
+		// The transaction doesn't exist yet on disk; use the supplied version.
+		reply.Txn = &txn
+	}
+
+	for i := range args.Intents {
+		// Set the transaction into the intents (TxnCoordSender avoids sending
+		// all of that redundant info over the wire).
+		// This needs to be done before the commit status check because
+		// the TransactionAbortedError branch below returns these intents.
+		args.Intents[i].Txn = *(reply.Txn)
 	}
 
 	// If the transaction record already exists, verify that we can either
 	// commit it or abort it (according to args.Commit), and also that the
 	// Timestamp and Epoch have not suffered regression.
 	if ok {
-		if existTxn.Status == proto.COMMITTED {
-			return reply, proto.NewTransactionStatusError(existTxn, "already committed")
-		} else if existTxn.Status == proto.ABORTED {
-			return reply, proto.NewTransactionAbortedError(existTxn)
-		} else if txn.Epoch < existTxn.Epoch {
-			return reply, proto.NewTransactionStatusError(existTxn, fmt.Sprintf("epoch regression: %d", txn.Epoch))
-		} else if txn.Epoch == existTxn.Epoch && existTxn.Timestamp.Less(txn.OrigTimestamp) {
+		if reply.Txn.Status == proto.COMMITTED {
+			return reply, nil, proto.NewTransactionStatusError(reply.Txn, "already committed")
```

Comment:
if it's already committed, someone already sent an `EndTransaction` with a list of intents successfully, so all the machinery has run previously. For the regression errors, well they should simply never happen in practice. You can't actually ever resolve a non-local intent in that case until the transaction times out, at least not currently. So in those cases you'd want to limit yourself to local intents, which is exactly the part that we don't handle yet - returning a batch on error. I'm going to open an issue on this, it won't be hard to improve.